### PR TITLE
Deprecate the `sign` and `sign_for` APIs (RIPD-1649):

### PR DESCRIPTION
**Pull request by nbougalis**
In order to facilitate transaction signing, `rippled` offers the `sign` and `sign_for` commands, which, given a seed, can sign or multi-sign a transaction respectively. These commands are accessible from the command line, as well as over the WebSocket and RPC interfaces that `rippled` can be configured to provide.

These commands, unfortunately, have significant security implications:

  1. They require divulging an account's seed (commonly known as a "secret key") to the server.
  2. When executing these commands against remote servers, the seeds can be transported over clear-text links.
  3. When executing these commands over the command line, the account seed may be visible using common tools that show running processes and may potentially be inadvertently stored by system monitoring tools or facilities designed to maintain a history of previously typed commands.

While this commit cannot prevent users from issuing these commands to a server, whether locally or remotely, it does disable them by default.

Server operators that require the ability to sign can re-enable them by adding the following stanza to their configuration file:
```
[signing_support]
true
```

Ripple discourages server operators from doing so and advises against using these commands, which will be removed in a future release. If you rely on these commands for signing, please migrate to a standalone signing solution as soon as possible. One option is to use `ripple-lib`; documentation is
available at https://developers.ripple.com/rippleapi-reference.html#sign.

If the commands are administratively enabled, the server includes a warning on startup and adds a new field in the resulting JSON, informing the caller that the commands are deprecated and may become unavailable at any time.

Acknowledgements:
Jesper Wallin for reporting this issue to Ripple.

Bug Bounties and Responsible Disclosures:
We welcome reviews of the rippled code and urge researchers to responsibly disclose any issues that they may find. For more on Ripple's Bug Bounty program, please visit: https://ripple.com/bug-bounty

### CODE REVIEW

File: src/ripple/core/Config.h [hunk 1]
```diff
@@ -107,6 +107,14 @@ class Config : public BasicConfig
     */
     bool                        RUN_STANDALONE = false;
 
+    /** Determines if the server will sign a tx, given an account's secret seed.
+
+        In the past, this was allowed, but this functionality can have security
+        implications. The new default is to not allow this functionality, but
+        a config option is included to enable this.
+    */
+    bool signingEnabled_ = false;
```
Comment (mellery451) [inline src/ripple/core/Config.h]: all the other configurable bools are `public`, which does make it easier to manipulate the value for testing. I'm ok if you want to keep it private, but just noting the inconsistency with the other fields.

File: src/test/app/MultiSign_test.cpp [hunk 2]
```diff
@@ -190,7 +190,12 @@ class MultiSign_test : public beast::unit_test::suite
     test_enablement()
     {
         using namespace jtx;
-        Env env(*this, FeatureBitset{});
+        Env env(*this, envconfig([](std::unique_ptr<Config> cfg)
```
Comment (mellery451) [inline src/test/app/MultiSign_test.cpp]: for your consideration: here's a commit that uses an `envconfig` helper for flipping this setting:  https://github.com/mellery451/rippled/commit/d440918cb9ab71ba362377544a7a4ae59bfd806f

**Comment by seelabs** [comment]
MultiSign_test.cpp and JSONRPC_test.cpp need to: `#include <ripple/core/ConfigSections.h>` or the nounity build breaks. Here's a patch: https://github.com/seelabs/rippled/commit/4933256a5bd2bb455296f264ea4fe95b9a92a0f7

**Comment by wilsonianb** [comment]
It looks like this doesn't cover `submit`'s Sign-and-submit mode.

What if instead of the config option, we just make `sign`, `sign_for`, and sign-and-`submit` admin-only?

**Comment by mDuo13** [comment]
The sign-and-submit mode of `submit` should definitely match what we do to `sign`. It's even more risky since you're not even looking at the signed result before submitting it!

Also, I like @wilsonianb's suggestion of making all of them admin-only rather than strictly removing them, because that way:

1. It matches the treatment of `wallet_propose`
2. It doesn't require changes from those who are using `rippled`'s signing securely using a local server
3. It still prevents the most insecure usage

I would be in favor of still disabling the commandline versions for the reasons described in the commit while leaving the WS / JSON-RPC versions usable on admin commands by default.

**Comment by MarkusTeufelberger** [comment]
Command-line usage still means that you need to be capable of installing, configuring and running rippled instead of just randomly submitting to some webservice endpoint online...

**Comment by nbougalis** [comment]
@seelabs: I think that deprecating these commands is sufficient for 1.1.0; my preference would be to _entirely_ remove them, leaving behind only a new `submit_tx` command that requires a signed blob, and a new tool that can be used to sign commands and produce the blob, or optionally, submit the signed blob to a server of your choice.

**Comment by seelabs** [comment]
@nbougalis I'm fine with that. :+1:

File: src/ripple/rpc/handlers/SignFor.cpp [hunk 3]
```diff
@@ -33,6 +33,12 @@ namespace ripple {
 // }
 Json::Value doSignFor (RPC::Context& context)
 {
+    if (!context.app.config().canSign())
```
[the remark this answers is not in the document]
Comment (nbougalis) [reply src/ripple/rpc/handlers/SignFor.cpp]: It is in the the PR; see https://github.com/ripple/rippled/blob/38c3a46a337916216a1e3ad481229fe87e07d542/src/ripple/rpc/handlers/SignFor.cpp#L36. I didn't push that branch up.

File: cfg/rippled-example.cfg [hunk 4]
```diff
@@ -1002,7 +1002,32 @@
...
 #
-# 8. Example Settings
+# 8. Misc Settings
+#
+#----------
+#
+# [signing_support]
+#
+#   Specifies whether the server will accept "sign" and "sign_for" commands
+#   from remote users. Even if the commands are sent over a secure protocol
```
[the remark this answers is not in the document]
Comment (nbougalis) [reply cfg/rippled-example.cfg]: Hmm, yes, that should porbably be changed.

File: src/ripple/app/main/Application.cpp [hunk 5]
```diff
@@ -1399,6 +1399,26 @@ bool ApplicationImp::setup()
...
+        JLOG(m_journal.warn()) <<
+            "*** commands. These commands have security implications and have";
+        JLOG(m_journal.warn()) <<
+            "*** been deprecated. They will be removed in a future release of";
+        JLOG(m_journal.warn()) <<
+            "*** rippled.";
+        JLOG(m_journal.warn()) <<
+            "*** If you do not use them to sign transactions please edit your";
+        JLOG(m_journal.warn()) <<
+            "*** configuration file and remove the [enable_signing] stanza.";
```
[the remark this answers is not in the document]
Comment (nbougalis) [reply src/ripple/app/main/Application.cpp]: The message will _only_ be displayed if the specified entry exists in the config file, therefore instructing the users to remove it.

**Comment by nbougalis** [comment]
Restricting the commands to admin-only and not providing a temporary way for server operators to revert this change is a breaking change. It breaks unit tests, which we could fix, but it may also break legitimate use cases, where granting administrative level access to the server is not plausible or advisable.

By having a config option, we effectively afford people a grace period (say, until 1.3,) during which time operators affected by this change can temporarily "revert" it until they migrate to a better solution.

File: src/ripple/app/main/Application.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (wilsonianb) [reply src/ripple/app/main/Application.cpp]: :man_facepalming:

File: cfg/rippled-example.cfg (hunk 4 above)
[the remark this answers is not in the document]
Comment (wilsonianb) [reply cfg/rippled-example.cfg]: Actually, I think it's fine.
I was thinking the default was no signing support and setting this field allowed just admin-only.

**Comment by MarkusTeufelberger** [comment]
https://developers.ripple.com/wallet_propose.html and https://developers.ripple.com/validation_create.html have the same "issues" as the sign methods, will they be deprecated too?

I disagree that `ripple-lib` is a viable alternative for creating a signed transaction, especially considering that it is written in JavaScript (or Typescript). The current `sign` methods are for example the only way to sign a transaction using Python. If they get deprecated it is basically impossible to write wallet software in anything but JavaScript without first taking a while to re-implement transaction serialization and testing it. Serialization is mostly documented in code, https://developers.ripple.com/transaction-formats.html does not explain how to serialize the various fields in a transaction to generate the correct hash/"index" for it. Compare this effort to making a local API call on an offline, trusted machine.

About submitting a signed expanded JSON object: To be able to generate a valid signature, one anyways needs a proper implementation of the way transactions are being serialized (something that's not readily available in most languages), so I'm not sure if that's a common use case.

TL;DR: Please do NOT deprecate this functionality as a whole, admin-only by default and public only via a switch (with that switch being potentially deprecated) is enough imho.

**Comment by nbougalis** [comment]
@MarkusTeufelberger: I'd much prefer a standalone signing tool, that does the signing and optionally can submit the result to a server, as opposed to keeping these commands; beyond the security implications, they are awkward to use and clutter the `rippled` interface; more generally, I'd prefer to reduce the code that is in `rippled` and would rather migrate a lot of the "client" functionality that it implements in separate binaries, including an RPC front-end tool.

Anyways, that's a different topic.

Migrating the commands over to admin-only (which we are doing with this) is fine for now and we can remove the the switch in a later release, preferably when tools become available.

**Comment by nbougalis** [comment]
Merged as 38c3a46a337916216a1e3ad481229fe87e07d542

**Comment by gituser** [comment]
Just stumbled upon this..

Until there is an example implementation to sign in all popular languages, e.g. go/php/python/c++/etc there is no point in removing this functionality.
Also as a node operator I completely agree with what @MarkusTeufelberger said earlier - make `sign`, `submit`, `sign-and-submit` admin-only and do not remove them completely and add configuration option to make them available for user endpoints.

**Comment by nbougalis** [comment]
The commands are now admin-only by default. They will be removed, but only after alternatives are available.

**Comment by passionofvc** [comment]
I have a question if sign remove in the future, how to sign tx if client use C#, java, python, or any other language rather than ripple-lib(javascript)?  Just make rippled offline completely, not connect to anywhere and only for local sign,and some base key/wallet functionality is sufficient I think, we can not stop  other people to using ripple-lib if he misused  in online pc.
